### ISSUE TITLE
Swapping in battle doesn't work

### ISSUE THREAD

**Issue by MirceaKitsune** [Tuxemon/Tuxemon#266]
Now that capturing other monsters is finally implemented, I also tested in-battle swapping. Unfortunately it seems to be completely broken and has some strange side effects.

Whenever I select Tuxemon in the middle of a battle and choose to swap to another monster, the player basically attacks their own Tuxemon with a supposed swap attack! I get the message "Player used swap!" while the attack sound plays, and the first time I even see the teeth animation over my own Tuxemon plus them flashing as if they took damage! The swapping itself never occurs, and you're still stuck with the Tuxemon you started the battle with.

There is however an odd exception: If you have more than one Tuxemon, and the one currently in battle faints, you can swap to the other one without any issues. The mentioned problem only appears to occur if you choose to swap of your own accord, instead of attacking or running or using an item.

**Comment by xirsoi** [comment]
I'd like to take a crack at this, if no one minds.
I've found the methods that allow swapping, and I think I can figure out why the menu part doesn't work.

**Comment by MirceaKitsune** [comment]
Sure as far as I go: I'm only familiar enough with the code to play with simple things, and usually get more motivated by working on features over bugfixes. Thanks!

**Comment by ShadowApex** [comment]
Right now when you select the "TUXEMON" menu item, it will call the `open_swap_menu` method in `combat_menus.py`. This method will basically push the "MonsterMenuState" into the currently active state and set up a function to be executed (called `swap_it`) when a monster is selected. The `swap_it` function will create a new "swap" technique, which right now doesn't swap the active monster.

Instead of creating a "swap" technique, you could alternatively look into making `open_swap_menu` call the `ask_player_for_monster` method inside the "CombatState":

```python
        combat_state = self.game.get_state_name("CombatState")
        for player in combat_state.human_players:
            combat_state.ask_player_for_monster(player)
```

The `ask_player_for_monster` method will not clear out the existing monster from play, though, so some additional logic might be needed to make it operate cleanly.

@bitcraft Any suggestions on how to handle switching monsters that haven't fainted?

**Comment by xirsoi** [comment]
Figured it out.

`ask_player_for_monster' calls 'add_monster_into_play`

which does what it says on the tin. I think that should fix it! I'll make the change, do a couple quick tests, and have a PR up tonight.

**Comment by xirsoi** [comment]
I was wrong, slightly more complicated. I'm pretty sure I can figure this out, it'll just take more time. Unfortunately I have little free time, so this could take me a week or two. I still wanna do it though!

**Comment by MirceaKitsune** [comment]
From what I'm seeing, some things in Tuxemon took months or years to do. So definitely nor a problem by me! Though this is a nasty and obvious bug, so just hope you get around to doing it is all.

**Comment by bitcraft** [comment]
I will fix this this week.

**Comment by xirsoi** [comment]
I'm sorry I didn't get this done. It proved more than I could do but was unwilling to admit defeat :(

**Comment by bitcraft** [comment]
No worries. There are a lot of moving parts in the combat class.  It is not
my cleanest implementation.  I will get this working by adding a function
to remove a monster in play with fainting it, then queue up an action to
bring in a new monster.   You were on the right track, but the key point I
think is the Combat.monsters_in_play dictionary that needs to be updated as
well.

**Comment by MirceaKitsune** [comment]
@bitcraft Will that approach be compatible with the potential implementation of items that revive Tuxemon in-battle? Unregistering a Tuxemon from existence in the battle system sounds like it would break that, so just something to point out.

**Comment by bitcraft** [comment]
No, there will be no conflicts with existing functionality.

**Comment by bitcraft** [comment]
Swapping works.  Question: should the opponents still get a turn that round?  For example, should swapping just mean the round restarts, or should the opponents get to attack you after the swap?

**Comment by xirsoi** [comment]
In Pokemon swapping counts as an action, so the opponent would then be able to make an attack. Assuming, of course, that swapping is faster than the enemy tuxemon.

This seems fine to me.

**Comment by bitcraft** [comment]
Cool.  Sounds good to me.  The fact that the swap counts as a turn is why
swap was implemented as a technique.  I'll have a PR for this tonight.
Thanks!

**Comment by bitcraft** [comment]
I have a quick question for the gamers.  When a monster is removed from play, but is the target of a technique, how should the action be handled?  It comes to two options, that I can tell:

1. Keep the original action, but change the target to the swapped monster
2. Give the AI (or player, for MP) the option to select a new action

I'm going to say #2 seems the most fair, because you could abuse the swaps.  I'm inventing a situation, so bear with me....  Suppose you have a weak monster and you enter battle.  You entice your opponent to hit you with a fire spell, so you swap to a monster who is healed by fire spells.  That to me seems unfair, but idk.

But abusing swaps could become a legit strategy, and I'm not against that.

TBH, I feel like I'm stressing about minutia, since the current AI is terribad and MP combat doesn't work, but I think its important to figure it out now, before it becomes a huge issue later. 

TLDR; Do we give the AI/opponent another chance to pick an action when a swap occurs, or does the action carry out, but targets the new monster?

Thoughts?

**Comment by Sanglorian** [comment]
> In Pokemon swapping counts as an action, so the opponent would then be able to make an attack. Assuming, of course, that swapping is faster than the enemy tuxemon.

>The fact that the swap counts as a turn is why swap was implemented as a technique. 

Just to clarify this, in Pokemon swapping takes place before any actions, regardless of the Speed of the Pokemon. So even if it would go (1) ROCKITTEN uses Rock Smash then (2) BUDAYE uses Sapling Strike, if BUDAYE swaps out, that happens before ROCKITTEN's Rock Smash. 

> I'm going to say #2 seems the most fair, because you could abuse the swaps. I'm inventing a situation, so bear with me.... Suppose you have a weak monster and you enter battle. You entice your opponent to hit you with a fire spell, so you swap to a monster who is healed by fire spells. That to me seems unfair, but idk.
> But abusing swaps could become a legit strategy, and I'm not against that.

I think "abusing" swaps is great! If you can anticipate your opponent's next move, you can mitigate it, whether that's by swapping in something that resists Fire when you think a Fire attack is coming, or whatever. That's good, tactical play.

Of course it does depend on swapping costing an action, otherwise it'd be too tempting to do.

**Comment by bitcraft** [comment]
So far, I've gotten two votes for #1: keep original action.  I'll go ahead and do it that way.  Thanks!

**Comment by xirsoi** [comment]
That's how I would do it too.

**Comment by bitcraft** [comment]
OK......  so no speed consideration of the monster/trainer?  Just always do a swap 1st?
